src/core/shutdown.c

```diff
@@ -159,6 +163,78 @@ static int switch_root_initramfs(void) {
         return switch_root("/run/initramfs", "/oldroot", false, MS_BIND);
 }
 
+static bool sync_making_progress(unsigned long long *prev_dirty) {
+        _cleanup_free_ char *dirty_string = NULL;
+        unsigned long long dirty = 0;
+        int r;
+
+        r = get_proc_field("/proc/meminfo", "Dirty", WHITESPACE, &dirty_string);
```

Comment:
When writing data out, pages normally go through 2 stages.  Firstly they are "Dirty" and sitting idly in the cache, then the writeback thread, or a "sync" or similar finds then, moves them to "Writeback" and submits them to the backend device.  When the backend device reports that the write is complete, they are done - except with NFS.  For NFS there is an extra stage represented by the COMMIT NFS operation (which effectively causes a "sync" on the server).  So for NFS, pages transition from Writeback to NFS_Unstable, and are not done until after the COMMIT completes.  So this code should really sample "NFS_Unstable", "Writeback" and "Dirty" (ideally in that order to avoid missing pages that change state) and use the sum.

Reply:
Sorry for the delay,

That sounds fair enough. I will include it in the next push.